File: nodejs_backend/src/RobotManager.js
/**
 * Robot Manager for ESP32 robot lifecycle and communication management.
 */

const Robot = require('./Robot');
const NetworkManager = require('./NetworkManager');

class RobotManager {
    static ROBOT_TIMEOUT_SECONDS = 60.0; // Mark robot as disconnected after 60 seconds (matches Python behavior of no timeout)
    static DISCOVERY_PORT = NetworkManager.DISCOVERY_PORT; // Import from NetworkManager

    constructor(networkManager, apiServer = null) {
        this.networkManager = networkManager;
        this.apiServer = apiServer;
        this.connectedRobots = new Map();
        this.discoveredRobots = new Map();
        this.currentGameState = "standby"; // "standby" or "teleop"
        this.emergencyStopActive = false;
        this.running = false;
        this.discoveryInterval = null;
        this.timeoutInterval = null;
        this.gameStatusInterval = null;
        this.disconnectCallbacks = [];

        console.log('[RobotManager] ESP32 Robot Manager initialized with discovery protocol');
    }

    registerDisconnectCallback(callback) {
        this.disconnectCallbacks.push(callback);
    }

    startDiscovery() {
        if (this.running) {
            console.warn('[RobotManager] Discovery already running');
            return;
        }

        this.running = true;

        // Listen for discovery messages (pong responses)
        this.networkManager.onDiscoveryMessage((message, rinfo) => {
            this._handleDiscoveryResponse(message, rinfo);
        });

        // Send discovery pings periodically
        this.discoveryInterval = setInterval(() => {
            this.networkManager.sendDiscoveryPing();
        }, 50); // Send ping every 50ms like the Python code

        // Start timeout check loop
        this.timeoutInterval = setInterval(() => {
            this._timeoutCheckLoop();
        }, 2000); // Check every 2 seconds

        // Start game status broadcast loop
        this.gameStatusInterval = setInterval(() => {
            this._broadcastGameStatus();
        }, 1000); // Broadcast every second like Python code

        console.log(`[RobotManager] Discovery service started - sending pings on port ${NetworkManager.ESP32_UDP_PORT}`);
    }

    _handleDiscoveryResponse(message, rinfo) {
        // Handle discovery pong from robot: "pong:<robotId>"
        if (message.startsWith("pong:")) {
            const robotId = message.substring(5).trim();
            const ipAddress = rinfo.address;

            // Check if robot is in connected list
            let robot = this.connectedRobots.get(robotId);
            if (robot) {
                // Update connected robot's last seen time
                robot.setIpAddress(ipAddress);
                robot.updateLastSeenTime();
            } else {
                // Add/update discovered robot
                robot = this.discoveredRobots.get(robotId);
                if (!robot) {
                    robot = new Robot(robotId, robotId, ipAddress, "discovered");
                    this.discoveredRobots.set(robotId, robot);
                    console.log(`[RobotManager] Discovered new robot: ${robotId} at ${ipAddress}`);
                } else {
                    robot.setIpAddress(ipAddress);
                    robot.updateLastSeenTime();
                }
            }
        }
    }

    getRobot(robotId) {
        let robot = this.connectedRobots.get(robotId);
        if (robot) return robot;
        return this.discoveredRobots.get(robotId);
    }

    getConnectedRobots() {
        return Array.from(this.connectedRobots.values());
    }

    getDiscoveredRobots() {
        return Array.from(this.discoveredRobots.values());
    }

    connectDiscoveredRobot(robotId) {
        const robot = this.discoveredRobots.get(robotId);
        if (robot) {
            // Move from discovered to connected
            this.discoveredRobots.delete(robotId);
            robot.setConnected(true);
            robot.status = "connected";
            this.connectedRobots.set(robotId, robot);
            console.log(`[RobotManager] Connected to robot: ${robotId}`);

            // Send current game status
            this.networkManager.sendGameStatus(robot.name, robot.ipAddress, this.currentGameState);

            return robot;
        }
        return null;
    }

    removeRobot(robotId) {
        if (this.connectedRobots.has(robotId)) {
            const robot = this.connectedRobots.get(robotId);
            // Send stop command before removing
            this.networkManager.sendRobotCommand(robot.name, robot.ipAddress, 127, 127, 127, 127);
            this.connectedRobots.delete(robotId);
            console.log(`[RobotManager] Removed robot: ${robotId}`);
        }
    }

    scanForRobots() {
        console.log('[RobotManager] Robot scan requested - discovery is automatic via broadcast');
    }

    startTeleop() {
        this.currentGameState = "teleop";
        console.log('[RobotManager] Teleop mode started');

        // Send teleop status to all connected robots
        for (const robot of this.connectedRobots.values()) {
            this.networkManager.sendGameStatus(robot.name, robot.ipAddress, "teleop");
        }
    }

    sendMovementCommand(robotId, leftX, leftY, rightX, rightY) {
        const robot = this.getRobot(robotId);
        if (!robot) {
            console.warn(`[RobotManager] Cannot send command - robot not found: ${robotId}`);
            return;
        }

        // Convert from -1.0 to 1.0 range to 0-255 range (127 = center)
        const leftXInt = Math.max(0, Math.min(255, Math.floor((leftX + 1.0) * 127.5)));
        const leftYInt = Math.max(0, Math.min(255, Math.floor((leftY + 1.0) * 127.5)));
        const rightXInt = Math.max(0, Math.min(255, Math.floor((rightX + 1.0) * 127.5)));
        const rightYInt = Math.max(0, Math.min(255, Math.floor((rightY + 1.0) * 127.5)));

        // Send command via network manager
        this.networkManager.sendRobotCommand(
            robot.name, robot.ipAddress,
            leftXInt, leftYInt, rightXInt, rightYInt
        );

        // Update last command time
        robot.updateLastCommandTime();

        // Broadcast that robot is receiving commands
        if (this.apiServer) {
            this.apiServer.broadcastRobotReceivingCommand(robotId, true);
        }
    }

    sendStopCommand(robotName) {
        const robot = this.getRobot(robotName);
        if (!robot) {
            console.warn(`[RobotManager] Robot not found: ${robotName}`);
            return;
        }

        // Send centered command (all axes at 127 = stopped)
        this.networkManager.sendRobotCommand(robot.name, robot.ipAddress, 127, 127, 127, 127);
        if (this.apiServer) {
            this.apiServer.broadcastRobotReceivingCommand(robot.id, false);
        }
    }

    _timeoutCheckLoop() {
        try {
            const currentTime = Date.now();
            const robotsToRemove = [];

            // Only remove from discovered list after very long timeout (for truly offline robots)
            // Connected robots are never auto-disconnected (matches Python behavior)
            for (const [robotId, robot] of this.discoveredRobots.entries()) {
                const timeSinceSeen = (currentTime - robot.lastSeenTime) / 1000;
                if (timeSinceSeen > RobotManager.ROBOT_TIMEOUT_SECONDS) {
                    console.log(`[RobotManager] Removing stale robot ${robotId} from discovered list (offline for ${timeSinceSeen.toFixed(1)}s)`);
                    this.discoveredRobots.delete(robotId);
                    robotsToRemove.push(robotId);
                }
            }
        } catch (error) {
            console.error(`[RobotManager] Error in timeout check loop: ${error.message}`);
        }
    }

    _broadcastGameStatus() {
        // Send game status to all connected robots periodically
        for (const robot of this.connectedRobots.values()) {
            this.networkManager.sendGameStatus(robot.name, robot.ipAddress, this.currentGameState);
        }
    }

    stopTeleop() {
        this.currentGameState = "standby";
        console.log('[RobotManager] Teleop mode stopped');

        // Send standby status to all connected robots
        for (const robot of this.connectedRobots.values()) {
            this.networkManager.sendGameStatus(robot.name, robot.ipAddress, "standby");
            this.sendStopCommand(robot.id);
            if (this.apiServer) {
                this.apiServer.broadcastRobotReceivingCommand(robot.id, false);
            }
        }
    }

    emergencyStopAll() {
        this.emergencyStopActive = true;
        this.networkManager.broadcastEmergencyStop(true);
        console.warn('[RobotManager] Emergency stop activated for all robots');
        if (this.apiServer) {
            for (const robotId of this.connectedRobots.keys()) {
                this.apiServer.broadcastRobotReceivingCommand(robotId, false);
            }
        }
    }

    deactivateEmergencyStop() {
        this.emergencyStopActive = false;
        this.networkManager.broadcastEmergencyStop(false);
        console.log('[RobotManager] Emergency stop deactivated');
    }

    shutdown() {
        this.running = false;
        if (this.discoveryInterval) {
            clearInterval(this.discoveryInterval);
        }
        if (this.timeoutInterval) {
            clearInterval(this.timeoutInterval);
        }
        if (this.gameStatusInterval) {
            clearInterval(this.gameStatusInterval);
        }

        // Send stop commands to all robots
        for (const robot of this.connectedRobots.values()) {
            this.sendStopCommand(robot.id);
        }

        console.log('[RobotManager] Robot manager shutdown complete');
    }
}

module.exports = RobotManager;

File: python_backend/robot_manager.py
"""
Robot Manager for ESP32 robot lifecycle and communication management.
"""

import logging
import threading
import time
from typing import Dict, List, Optional
from robot import Robot, ESP32Command
from network_manager import NetworkManager

logger = logging.getLogger(__name__)


class RobotManager:
    """Manages ESP32 robot discovery, connection, and command transmission"""

    ROBOT_TIMEOUT_SECONDS = 60.0  # Mark robot as disconnected after 60 seconds of no pings

    def __init__(self, network_manager: NetworkManager, api_server=None): # Add api_server parameter
        self.network_manager = network_manager
        self.api_server = api_server # Store api_server instance
        self.connected_robots: Dict[str, Robot] = {}
        self.discovered_robots: Dict[str, Robot] = {}
        self.current_game_state = "standby"  # "standby" or "teleop"
        self.emergency_stop_active = False
        self._discovery_thread: Optional[threading.Thread] = None
        self._timeout_thread: Optional[threading.Thread] = None
        self._running = False
        self._lock = threading.Lock()
        self._disconnect_callbacks = []

        logger.info("ESP32 Robot Manager initialized with discovery protocol")
    
    def register_disconnect_callback(self, callback):
        """Register a callback to be called when a robot disconnects"""
        self._disconnect_callbacks.append(callback)

    def start_discovery(self):
        """Start discovery service - listens for robot discovery pings"""
        if self._running:
            logger.warn("Discovery already running")
            return

        self._running = True
        self._discovery_thread = threading.Thread(target=self._discovery_loop, daemon=True)
        self._discovery_thread.start()
        self._timeout_thread = threading.Thread(target=self._timeout_check_loop, daemon=True)
        self._timeout_thread.start()
        logger.info(f"Discovery service started on port {NetworkManager.DISCOVERY_PORT}")
    
    def _discovery_loop(self):
        """Main discovery loop - listens for robot pings"""
        while self._running:
            try:
                message = self.network_manager.receive_discovery_message()
                if message and message.startswith("DISCOVER:"):
                    self._handle_discovery_ping(message)
            except Exception as e:
                logger.debug(f"Discovery listening error: {e}")
            
            time.sleep(0.5)  # Check every 500ms
    
    def _handle_discovery_ping(self, message: str):
        """
        Handle discovery ping from robot: "DISCOVER:<robotId>:<IP>"
        """
        parts = message.split(':')
        if len(parts) >= 3:
            robot_id = parts[1]
            ip_address = parts[2]

            with self._lock:
                # Check if robot is in connected list
                robot = self.connected_robots.get(robot_id)
                if robot:
                    # Update connected robot's last seen time
                    robot.set_ip_address(ip_address)
                    robot.update_last_seen_time()
                else:
                    # Add/update discovered robot
                    robot = self.discovered_robots.get(robot_id)
                    if robot is None:
                        robot = Robot(robot_id, robot_id, ip_address, "discovered")
                        self.discovered_robots[robot_id] = robot
                        logger.info(f"Discovered new robot: {robot_id} at {ip_address}")
                    else:
                        robot.set_ip_address(ip_address)
                        robot.update_last_seen_time()



    def send_stop_command(self, robot_name: str):
        """Send stop command to specific robot"""
        robot = self.get_robot(robot_name)
        if not robot:
            logger.warn(f"Robot not found: {robot_name}")
            return

        stop_command = ESP32Command.create_stop_command(robot_name)
        self._send_esp32_command(robot, stop_command)
        if self.api_server:
            self.api_server.broadcast_robot_receiving_command(robot.id, False)

    def _timeout_check_loop(self):
        """Check for robots that have timed out and mark them as disconnected"""
        while self._running:
            try:
                current_time = time.time()
                disconnected_robots = []
                robots_to_remove = []

                with self._lock:
                    # Check connected robots for timeout
                    for robot_id, robot in list(self.connected_robots.items()):
                        time_since_seen = current_time - robot.last_seen_time
                        if time_since_seen > self.ROBOT_TIMEOUT_SECONDS:
                            logger.warn(f"Robot {robot_id} timed out (no ping for {time_since_seen:.1f}s). Paired controller: {robot.paired_controller_id}")
                            # Broadcast receiving: False before removing
                            if self.api_server:
                                self.api_server.broadcast_robot_receiving_command(robot_id, False)
                            # Remove immediately from connected list - don't keep disconnected robots
                            del self.connected_robots[robot_id]
                            disconnected_robots.append(robot_id)
                            robots_to_remove.append(robot_id)

                    # Also remove from discovered list - robots should completely disappear after timeout
                    for robot_id, robot in list(self.discovered_robots.items()):
                        time_since_seen = current_time - robot.last_seen_time
                        if time_since_seen > self.ROBOT_TIMEOUT_SECONDS:
                            logger.info(f"Removing stale robot {robot_id} from discovered list (offline for {time_since_seen:.1f}s)")
                            del self.discovered_robots[robot_id]
                            if robot_id not in robots_to_remove:
                                robots_to_remove.append(robot_id)

                # Call disconnect callbacks outside of lock to avoid deadlock
                for robot_id in disconnected_robots:
                    for callback in self._disconnect_callbacks:
                        try:
                            callback(robot_id)
                        except Exception as e:
                            logger.error(f"Error in disconnect callback: {e}")

            except Exception as e:
                logger.error(f"Error in timeout check loop: {e}")

            time.sleep(2.0)  # Check every 2 seconds

    def stop_teleop(self):
        """Disable teleop mode - robots stop moving"""
        self.current_game_state = "standby"
        logger.info("Teleop mode stopped")

        # Send standby status to all connected robots
        with self._lock:
            for robot in self.connected_robots.values():
                self.network_manager.send_game_status(robot.name, robot.ip_address, "standby")
                self.send_stop_command(robot.id)
                if self.api_server:
                    self.api_server.broadcast_robot_receiving_command(robot.id, False)

    def emergency_stop_all(self):
        """Activate emergency stop for all robots"""
        self.emergency_stop_active = True
        self.network_manager.broadcast_emergency_stop(activate=True)
        logger.warn("Emergency stop activated for all robots")
        if self.api_server:
            for robot_id in self.connected_robots.keys():
                self.api_server.broadcast_robot_receiving_command(robot_id, False)

    def deactivate_emergency_stop(self):
        """Deactivate emergency stop"""
        self.emergency_stop_active = False
        self.network_manager.broadcast_emergency_stop(activate=False)
        logger.info("Emergency stop deactivated")
        # No need to broadcast receiving: True here, as robots will only start receiving
        # commands again if a controller is actively sending them.

    
    def shutdown(self):
        """Shutdown robot manager"""
        self._running = False
        if self._discovery_thread:
            self._discovery_thread.join(timeout=2)
        
        # Send stop commands to all robots
        with self._lock:
            for robot in self.connected_robots.values():
                self.send_stop_command(robot.id)
        
        logger.info("Robot manager shutdown complete")
